diff --git a/CMakeLists.txt b/CMakeLists.txt
index c99c1ee..a546baa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -38,9 +38,22 @@ set (SOURCES ${SRC_DIR}/vland.c ${SRC_DIR}/vland_ovsdb_if.c)
 # Rules to build ops-vland
 add_executable (${VLAND} ${SOURCES})
 
-target_link_libraries (${VLAND} ${OVSCOMMON_LIBRARIES} ${OVSDB_LIBRARIES}
+target_link_libraries (${VLAND} ${OVSCOMMON_LDFLAGS} ${OVSDB_LDFLAGS}
                        -lpthread -lrt)
 
 # Rules to install ops-vland binary in rootfs
 install(TARGETS ${VLAND}
         RUNTIME DESTINATION bin)
+
+###
+### Packaging
+###
+SET(CPACK_DEBIAN_PACKAGE_NAME "${VLAND}")
+SET(CPACK_DEBIAN_PACKAGE_VERSION "0.1.0")
+SET(CPACK_DEBIAN_PACKAGE_ARCH "amd64")
+SET(CPACK_DEBIAN_PACKAGE_DEPENDS "libc6, openvswitch-common")
+SET(CPACK_DEBIAN_PACKAGE_MAINTAINER "OpsDev <ops-dev@lists.openswitch.net>")
+SET(CPACK_DEBIAN_PACKAGE_DESCRIPTION_SUMMARY "OpenSwitch VLAN daemon.")
+SET(CPACK_DEBIAN_PACKAGE_DESCRIPTION "Controls VLANs.")
+SET(CPACK_GENERATOR "DEB")
+INCLUDE(CPack)
