diff --git a/CMakeLists.txt b/CMakeLists.txt
index 82cb6b0..5d48a87 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -37,11 +37,24 @@ set (SOURCES ${SRC_DIR}/arpmgrd.c)
 
 # Rules to build arpmgrd
 add_executable (${ARPMGRD} ${SOURCES})
-target_link_libraries (${ARPMGRD} ${OVSCOMMON_LIBRARIES}
-                       ${OVSDB_LIBRARIES}
+target_link_libraries (${ARPMGRD} ${OVSCOMMON_LDFLAGS}
+                       ${OVSDB_LDFLAGS}
                        -lpthread -lrt)
 
 # Rules to install arpmgrd binary in rootfs
 install(TARGETS ${ARPMGRD}
     RUNTIME DESTINATION bin)
 
+
+###
+### Packaging
+###
+SET(CPACK_DEBIAN_PACKAGE_NAME "${ARPMGRD}")
+SET(CPACK_DEBIAN_PACKAGE_VERSION "0.1.0")
+SET(CPACK_DEBIAN_PACKAGE_ARCH "amd64")
+SET(CPACK_DEBIAN_PACKAGE_DEPENDS "libc6, libconfig-yaml, openvswitch-common")
+SET(CPACK_DEBIAN_PACKAGE_MAINTAINER "OpsDev <ops-dev@lists.openswitch.net>")
+SET(CPACK_DEBIAN_PACKAGE_DESCRIPTION_SUMMARY "OpenSwitch ARP daemon.")
+SET(CPACK_DEBIAN_PACKAGE_DESCRIPTION "Updates and manages the neighbor entries learned by the kernel in the OpenSwitch Neighbor table.")
+SET(CPACK_GENERATOR "DEB")
+INCLUDE(CPack)
